## Further Step Check List:
- 重新设计MongoDB的Database，对restaurant info和restaurant review分别存储
- 把目前的IO代码，直接改成从MongoDB中提取
- 设计一个函数，进行分段爬取。（因为不知道yelp对每个ip单位时间的限制是多少次，每天坚持爬一点就可以）
- 疑惑三连：MongoDB可以每个_id存多条数据吗？它是unique的“主键”吗？？？？？？？？？？？？？
    - 好像一个id只能保留一条记录，要修改data format，一个id对应爬下来的所有reviews

In [1]:
# Setup Libraries
import os
import re
import time
import string
import pandas as pd
import numpy as np
from itertools import product

import requests
import json
from bs4 import BeautifulSoup
from lxml import etree

# connect to MongoDB
import pymongo
from pymongo import MongoClient

import matplotlib.pyplot as plt
import seaborn as sns

In [131]:
from selenium import webdriver
from selenium.webdriver.common.keys import Keys    # type things in the search bar and see the results

from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver import ActionChains
from selenium.webdriver.support import expected_conditions as EC

import time
from bs4 import BeautifulSoup
import time
from lxml import etree

PATH = '/Users/liqingran/Downloads/chromedriver'

from selenium.webdriver.chrome.options import Options   # 实现无可视化界面
from selenium.webdriver import ChromeOptions            # 实现selenium规避检测

# 实例化options并进行参数设定，实现无可视化界面
chrome_options = Options()
chrome_options.add_argument('--headless')
chrome_options.add_argument('--disable-gpu')

# # 实现检测规避，这个不懂怎么加，之后再说  
# option = ChromeOptions()
# option.add_experimental_option('excludeSwitches', ['enable-automation'])
# driver = Chrome(options=option)

In [261]:
# my connection string into your application code
connect = 'mongodb+srv://m001-student:m001-mongodb-basics@sandbox.jqgjp.mongodb.net/restaurant_info?retryWrites=true&w=majority'
cluster = MongoClient(connect)
db = cluster['yelp_dataset']               # to database
restaurant_collection = db['restaurants']  # to collection
review_collection = db['reviews']             

### 1 Import restaurants info json file

Note: this file is directly imported from yelp API.

Further: Connect with MongoDB.

#### Key festures:
- _id: Yelp给每个business都生成了一个unique id。NoSQL数据库就不用自动生成id了。
- name
- phone
- price
- rating
- url
- address
- category
- transactions
- total review counts

#### 1.1 Import from local file ====This step is not necessary====
In the circumstance that the `.json` file is not large, using `json.loads()` would be good enough.

If the file is too large (up to several GB), using Spark might be a smarter idea.

Since the restaurants info for one single city only takes up 500KB storage. I then use `json.loads`.

In [3]:
with open("nyc_restaurants.json", 'r') as f:
    nyc = json.loads(f.read())

In [4]:
nyc

[{'_id': 'fVbUVAiLiGgLA_nxBFxyww',
  'name': 'Thursday Kitchen',
  'phone': '',
  'price': '$$',
  'rating': 4.5,
  'url': 'https://www.yelp.com/biz/thursday-kitchen-new-york?adjust_creative=b1l9PtKsY5nsB8qEsNcf0Q&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=b1l9PtKsY5nsB8qEsNcf0Q',
  'address': '424 E 9th St New York, NY 10009',
  'category': ['Korean', 'American (New)', 'Tapas/Small Plates'],
  'transactions': ['pickup', 'delivery'],
  'review_count': 1373},
 {'_id': 'ETgJqJHV7BW6pIr9Ox74sA',
  'name': 'Amélie',
  'phone': '(212) 533-2962',
  'price': '$$',
  'rating': 4.5,
  'url': 'https://www.yelp.com/biz/am%C3%A9lie-new-york?adjust_creative=b1l9PtKsY5nsB8qEsNcf0Q&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=b1l9PtKsY5nsB8qEsNcf0Q',
  'address': '22 W 8th St New York, NY 10011',
  'category': ['French', 'Wine Bars'],
  'transactions': ['pickup', 'delivery'],
  'review_count': 2708},
 {'_id': 'q11TljTQd33XCWlVoPyRqg',
  'name': 'The

#### 1.2 Load data from MongoDB database

In [276]:
restaurants_result = restaurant_collection.find({})

### 2 Extract `_id` and `url`

This step is used for constructing reviews database. For that purpose, I only remain id and url.

In [324]:
nyc_basic_info = []
for res in restaurants_result:
    dict_ = {
        '_id': res['_id'],
        'url': res['url'],
        'name': res['name']
    }
    nyc_basic_info.append(dict_)

### 3 Constructing `review` database

#### 3.1 Functions
The following functions are used for scraping reviews using each restaurants' urls.

- `organize_info`: A utility function that clean the format of each records (ie. turning list of lists into list of dictionaries)
- `retrive_review`: Retrive real-time reviews for each source page.
- `get_max_page`: Get the total page number for further click iteration.
- `next_page_button`: Find the next page button for webdriver to click
- `scrap_review`: Scrap reviews given one restaurant url
- `get_all_reviews`: Scrap all reviews for input info. Returns dictionary-style of results, each result is a review record.

#### 3.2 Data format
The __review__ database has the following features:
- Name: (Not necessary to have this feature, I just scraped it for convinience. May delete this in the future)
- Rating: ratings on each review given by user
- Date: date published this review
- Review: the review text
- Image: image urls if this review contains image(s). If no image posted, then it's empty [''].
- business_id: id of this business. Used for link the restaurant database.

When loading into MongoDB, it will automatically generate a ObjectId `_id`. I then consider it as review_id. The links between `restaurants` and `reviews` tables are `restaurants._id` and `reviews.business_id`.

In [323]:
def get_max_page(parser):
    """Get the total page number for further click iteration."""
    num = parser.xpath('//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[2]/span/text()')
    num = int(num[0].split(' ')[-1])
    return num


def organize_info(info):
    """Since each time we return lists of dictionaries, 
       this function is for concatinating all dictionaries into one single list."""
    final_info = []
    for i in info:
        for j in i:
            final_info.append(j)
    return final_info


def next_page_button(parser):
    """Find the element of the clicking next page button."""
    button_xpath = '//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div'
    page_divs = parser.xpath(button_xpath)
    click_point = len(page_divs)
    click_xpath = button_xpath + str([click_point]) + '/span/a/span'
    return click_xpath

In [321]:
def retrive_review(parser):
    """Retrive real-time reviews for each source page."""
    info = []
    reviews = []
    ratings = []
    dates = []
    images = []
    name = parser.xpath('//*[@id="wrap"]/div[3]/yelp-react-root/div/div[2]/div[1]/div[1]/div/div/div[1]/h1/text()')
    li_list = parser.xpath('//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/ul/li')
    for li in li_list:
        rating = li.xpath('./div/div[2]/div/div[1]/span/div/@aria-label')
        date = li.xpath('./div/div[2]/div/div[2]/span/text()') 
        review = li.xpath('./div/div[3]/p/span/text()')  
        if review:    # 因为如果这个review没放图片就在div[3]，如果放了图片就在div[4]
            reviews.append(review[0])
            images.append([''])
        else:
            review = li.xpath('./div/div[4]/p/span/text()')
            reviews.append(review[0])
            img_divs = li.xpath('./div/div[5]/div')
            img_links = []
            for d in img_divs:
                img_link = d.xpath('./div/div/div/a/img/@src')
                img_links.append(img_link)
            images.append(img_links[0])
        
        ratings.append(rating[0])
        dates.append(date[0])
        
    for i in range(len(reviews)):
        dict_ = {
            "name": name[0],
            "rating": ratings[i],
            "date": dates[i],
            "review": reviews[i],
            "image": images[i]
        }
        
        info.append(dict_)        
    return info

In [257]:
def scrap_review(url):
#     PATH = '/Users/liqingran/Downloads/chromedriver'
    driver = webdriver.Chrome(executable_path = PATH, options = chrome_options)
#     driver = webdriver.Chrome(executable_path = PATH)
    
    res = driver.get(url)
    time.sleep(5)
    parser = etree.HTML(str(driver.page_source).replace('<br>', '\n').replace('</br>', '\n'))
    
    total_page_num = get_max_page(parser)   
    next_page_xpath = next_page_button(parser)

    info = []
    if total_page_num < 10:
        for i in range(total_page_num):
            if i != total_page_num-1:
                dict_ = retrive_review(parser)
                info.append(dict_)
#                 print("->Page", i+1, "sucess!") 
                element = driver.find_element_by_xpath(next_page_xpath)        
#                 webdriver.ActionChains(driver).move_to_element(element).click(element).perform()
                webdriver.ActionChains(driver).click(element).perform()
                time.sleep(5)
                parser = etree.HTML(str(driver.page_source).replace('<br>', '\n').replace('</br>', '\n'))
            else:
                dict_ = retrive_review(parser)
                info.append(dict_)
#                 print("->Page", i+1, "sucess!")
    else:
        for i in range(9):
            dict_ = retrive_review(parser)
            info.append(dict_)
#             print("->Page", i+1, "sucess!")
            element = driver.find_element_by_xpath(next_page_xpath)
            webdriver.ActionChains(driver).click(element).perform()
            time.sleep(5)
            parser = etree.HTML(str(driver.page_source).replace('<br>', '\n').replace('</br>', '\n'))

    driver.quit()
    return info

In [325]:
def get_all_reviews(restaurant_info):
    """
    Get all reviews from each restaurant in the restaurants list.
    Returns list of lists where each list is scraped reviews of one restaurant. Max 90 reviews for each restaurants.
    """
    final_reviews = []
    for item in restaurant_info:
        res_id = item['_id']
        res_url = item['url']
        try:
            info = scrap_review(res_url)
            info = organize_info(info)
            for i in info:
                i['business_id'] = res_id
            print(info[-1]['name'] + ' scrapped finished. ' + '\t\tTotal: ', str(len(info)) + ' reviews.')
        except Exception as e:
            print(item['name'] + ' scraping got exception. Exception ', e)
            continue
        
        final_reviews.append(info)
        time.sleep(10)
    return final_reviews

In [326]:
test1 = get_all_reviews(nyc_basic_info[100:105])

Shu Jiao Fu Zhou Cuisine Restaurant scrapped finished. 		Total:  90 reviews.
Saigon Social scrapped finished. 		Total:  75 reviews.
DOMODOMO - New York scrapped finished. 		Total:  90 reviews.
Hak Box got exception  Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[6]/span/a/span"}
  (Session info: headless chrome=89.0.4389.114)

Paesano of Mulberry Street scrapped finished. 		Total:  90 reviews.


test1返回一个list of lists。 每个list是一个restaurant爬下来的所有reviews（目前代码最多取90条，我觉得差不多够了）。

In [327]:
print(len(test1))
print(len(organize_info(test1)))

4
345


### 4 Store in MongoDB `reviews` database

In [254]:
test1 = organize_info(test1)
len(test1)

180

In [279]:
results = review_collection.delete_many({})

In [256]:
review_collection.insert_many(test1)

### 5 Scaping by parts 批量爬取

因为不知道yelp对于每个ip刷新网页的爬取限制，因此决定分段爬，这样能暂时规避ip被暂停的风险，也能在爬取中途出现报错时，快速定位到问题。

一些比较奇怪的餐厅记录(但应该问题不大)：
- 164. Barbalu Restaurant: 返回了100条reviews
- 221. La Sirène-Soho: 返回了160条reviews
- 298. Sweet Chick: 返回100条reviews


遇到错误的餐厅记录：
- 104. Hak Box
    - `Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[6]/span/a/span"}
  (Session info: headless chrome=89.0.4389.114)`
  
- 160. Bites of Xi'an
    - `Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[10]/span/a/span"}`
    
- 208. Mi Salsa Kitchen
    - `list index out of range` 这个估计是网页没加载好，就试图定位页数标签造成的。再爬一次就可以。

- 263. Landhaus
    - `list index out of range`
    
- 325. Ichibantei 
    - `list index out of range`
    
- 345. Berimbau Brazilian Kitchen
    - `Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[8]/span/a/span"}
  (Session info: headless chrome=89.0.4389.114)`
  
- 374. Madre
    - `list index out of range`
 
- 433. Da Gennaro
    - `list index out of range`
    
- ?671. Taqueria Fort Greene
    - `Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[8]/span/a/span"}
  (Session info: headless chrome=89.0.4389.114)`
  
- 657. Hi-Collar
    - `Message: disconnected: Unable to receive message from renderer
  (Session info: headless chrome=89.0.4389.114)`
  
- 680. Joor Chicken and Frozen Dessert
    - `list index out of range`
    
- 682. Four Four South Villag
    - `list index out of range`
    
- 821. Ming's Caffe
    - `Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[9]/span/a/span"}
  (Session info: headless chrome=89.0.4389.114)`
  
- 843. Navungerleider Sushi
    - `list index out of range`
    
- 865. Kjun
    - `list index out of range`
    
- 921. Coarse NY 
    - `Message: unknown error: net::ERR_ADDRESS_UNREACHABLE
  (Session info: headless chrome=89.0.4389.114)`
  
- 915. Cadence
    - `list index out of range`
    
- 936. Shinbashi 72
    - `Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[6]/span/a/span"}
  (Session info: headless chrome=89.0.4389.114)`
  
- 956. Bronx River Yacht Club
    - `list index out of range`
    
- 972. Amazon Go
    - `Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[5]/span/a/span"}
  (Session info: headless chrome=89.0.4389.114)`
  
- 987. The Lost Lady
    - ` Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[11]/span/a/span"}
  (Session info: headless chrome=89.0.4389.114)`

- 994. 最后一个没有爬

In [288]:
def part_scraping(restaurant_list, start_idx, end_idx):
    partial_results = get_all_reviews(restaurant_list[start_idx: end_idx])
    return partial_results

In [299]:
# results = collection.delete_many({})
def load_mongo(partial_results):
    res = organize_info(partial_results)
    review_collection.insert_many(res)

In [328]:
# temp_result = part_scraping(nyc_basic_info, 100, 120)
# load_mongo(temp_result)

Shu Jiao Fu Zhou Cuisine Restaurant scrapped finished. 		Total:  90 reviews.
Saigon Social scrapped finished. 		Total:  75 reviews.
DOMODOMO - New York scrapped finished. 		Total:  90 reviews.
Hak Box got exception  Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[6]/span/a/span"}
  (Session info: headless chrome=89.0.4389.114)

Paesano of Mulberry Street scrapped finished. 		Total:  90 reviews.
Peppa's Jerk Chicken scrapped finished. 		Total:  14 reviews.
Evil Katsu scrapped finished. 		Total:  20 reviews.
Soco scrapped finished. 		Total:  90 reviews.
hanon scrapped finished. 		Total:  90 reviews.
Saigon Shack scrapped finished. 		Total:  90 reviews.
Miss Ada scrapped finished. 		Total:  90 reviews.
The Smith scrapped finished. 		Total:  90 reviews.
3 Times 茅庐 scrapped finished. 		Total:  90 reviews.
Coco Bistro scrapped finish

In [332]:
# temp_result = part_scraping(nyc_basic_info, 150, 180)
# load_mongo(temp_result)

BoCaphe scrapped finished. 		Total:  90 reviews.
Boutros scrapped finished. 		Total:  90 reviews.
Loco Coco - Brooklyn scrapped finished. 		Total:  21 reviews.
Claw Daddy's scrapped finished. 		Total:  90 reviews.
Le Petit Monstre scrapped finished. 		Total:  14 reviews.
Banh Mi Saigon scrapped finished. 		Total:  90 reviews.
Les Enfants de Boheme scrapped finished. 		Total:  90 reviews.
Mekelburg's scrapped finished. 		Total:  60 reviews.
MONOMONO scrapped finished. 		Total:  90 reviews.
Bites of Xi'an got exception  Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/div/div[4]/div[1]/div/div[10]/span/a/span"}
  (Session info: headless chrome=89.0.4389.114)

Boqueria Soho scrapped finished. 		Total:  90 reviews.
Weekends scrapped finished. 		Total:  44 reviews.
Le Crocodile scrapped finished. 		Total:  65 reviews.
Barbalu Restaurant scrapped finished. 		T

In [358]:
# temp_result = part_scraping(nyc_basic_info, 330, 350)
# load_mongo(temp_result)

Paulie Gee's scrapped finished. 		Total:  90 reviews.
Fat Choy scrapped finished. 		Total:  31 reviews.
The Leroy House scrapped finished. 		Total:  90 reviews.
The Grey Dog - Nolita scrapped finished. 		Total:  90 reviews.
Gersi scrapped finished. 		Total:  90 reviews.
A-Pou's Taste scrapped finished. 		Total:  74 reviews.
Rice kitchen scrapped finished. 		Total:  59 reviews.
Dim Sum Palace scrapped finished. 		Total:  90 reviews.
Antica Ristorante scrapped finished. 		Total:  90 reviews.
We Rub You scrapped finished. 		Total:  4 reviews.
Jacob's Pickles scrapped finished. 		Total:  90 reviews.
Toro Loco scrapped finished. 		Total:  90 reviews.
Green Garden Village scrapped finished. 		Total:  60 reviews.
Gramercy Tavern scrapped finished. 		Total:  90 reviews.
Berimbau Brazilian Kitchen got exception  Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/div/div/div[2]/div/div[1]/div[2]/section[2]/div[2]/di

In [401]:
# temp_result = part_scraping(nyc_basic_info, 920, -1)
# load_mongo(temp_result)

Aqua Boil scrapped finished. 		Total:  90 reviews.
Zhongzhong Noodles scrapped finished. 		Total:  43 reviews.
Cornerstone Cafe scrapped finished. 		Total:  90 reviews.
Up Thai scrapped finished. 		Total:  90 reviews.
Calle Dao - Chelsea scrapped finished. 		Total:  90 reviews.
East Hae scrapped finished. 		Total:  36 reviews.
Jack's Wife Freda scrapped finished. 		Total:  90 reviews.
Cask scrapped finished. 		Total:  90 reviews.
Egg Shop scrapped finished. 		Total:  90 reviews.
Pinwheel Garden Dumpling & Noodle Bar scrapped finished. 		Total:  90 reviews.
Black Walnut Public House scrapped finished. 		Total:  39 reviews.
Thai Street Food scrapped finished. 		Total:  31 reviews.
Darna Falafel scrapped finished. 		Total:  90 reviews.
Don Ceviche scrapped finished. 		Total:  29 reviews.
ON scrapped finished. 		Total:  90 reviews.
Shinbashi 72 got exception  Message: no such element: Unable to locate element: {"method":"xpath","selector":"//*[@id="wrap"]/div[3]/yelp-react-root/div/div[3]/